rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md

```diff
@@ -0,0 +1,555 @@
+
+For example, in natural language processing, tokenization and vectorization may
+affect the later model type. They will need to find a way to pass this
+information from `Tuner.run_trial()` to HyperModel.build.
+
+#### Tune existing Keras code
+
+If the users have their code for model building and training ready written using
+Keras, and they want to tune some of the hyperparameters, they would have to
+change the code a lot to separate their code apart and wire the data flow and
+control flow between the overridden functions.
+
+#### Retrain the model after tuning
+
+If the user wants to retrain the model using the best hyperparameter values
+found, there is not a straight-forward way to do it if they used the
+hyperparameter in `Tuner.run_trial()` for data preprocessing and model
+training.
+
+## User Benefit
+
+The use cases described above would all have smooth workflows, without much
+extra code or learning of the framework.
+
+## Design Proposal
+
+We propose two workflows: the `Tuner` workflow and the `HyperModel` workflow to
+solve the problems above.
+
+The `Tuner` workflow is to override `Tuner.run_trial()`. The user can put all the
+code for data preprocessing, model building, model training all in one place in
+the `Tuner.run_trial()` function. No `HyperModel` is needed. It supports all the
+use cases mentioned above by providing the maximum freedom to the user.
+
+The `HyperModel` workflow follows the original `HyperModel` style. It is easier
+to learn and needs less code compared to the first workflow, but covers all the
+use cases as long as the code for building and training the model are separate.
+The user only needs to override the `HyperModel.fit()` for any tuning of the
+data preprocessing and model fitting process.
+
+## Detailed Design
+
+### The `Tuner` workflow
+
+Here is an end-to-end code example of the new workflow.
+
+The user only needs to override `Tuner.run_trial()` to put everything together,
+including data preprocessing, model building, and model training. It returns
+the evaluation results back to the tuner. 
+
+```py
+class MyTuner(kt.Tuner):
+  def run_trial(self, trial, x, y, callbacks=None, **kwargs):
+    hp = trial.hyperparameters
+    # Data preprocessing
+    num_features = hp.Int("num_features", 10, 15)
+    x, y = feature_selection(num_features=num_features, x, y)
+    # Model building
+    # Input shape depending on data preprocessing.
+    inputs = keras.Input(shape=(num_features,))
+    outputs = keras.layers.Dense(
+        hp.Choice('units', [8, 16, 32]),
+        activation='relu')(inputs)
+    outputs = keras.layers.Dense(1, activation='relu')(outputs)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    model.compile(loss='mse',
+                  metrics=['mae'])
+    # Model training
+    history = model.fit(
+        x,
+        y,
+        epochs=100,
+        validation_data=validation_data,
+        # Tune whether to use shuffle.
+        shuffle=hp.Boolean("shuffle"),
```

Comment:
What about `tf.data.Dataset` objects instead of numpy arrays? I suspect as long as `feature_selection()` yields the data in a format accepted by Keras, it should be fine? I am asking this irrespective of the `shuffle` argument. 

If so, I think it's worth specifying that.

Reply:
Yes, however, the `shuffle` arg only works for non-tf.data.Dataset type. For `tf.data.Dataset`, we can use `tf.data.Dataset.shuffle()`.

Reply:
Sure. Later in the examples, we have the NLP case using `tf.data`.

Reply:
Of course. Might be worth specifying this if not already done.